sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/BufferSendState.scala

```diff
@@ -0,0 +1,232 @@
+                hostBuffs += blockRange.rangeSize()
+            }
+            RangeBuffer(blockRange, rapidsBuffer)
+          }
+        }
+
+        logDebug(s"Occupancy for bounce buffer is [device=${deviceBuffs}, host=${hostBuffs}] Bytes")
+
+        bounceBuffToUse = if (deviceBuffs >= hostBuffs || hostBounceBuffer == null) {
+          deviceBounceBuffer.buffer
+        } else {
+          hostBounceBuffer.buffer
+        }
+
+        acquiredBuffs.foreach { case RangeBuffer(blockRange, rapidsBuffer) =>
+          require(blockRange.rangeSize() <= bounceBuffToUse.getLength - buffOffset)
+          withResource(rapidsBuffer.getMemoryBuffer) { memBuff =>
+            bounceBuffToUse match {
+              case _: HostMemoryBuffer =>
+                //TODO: HostMemoryBuffer needs the same functionality that
+                // DeviceMemoryBuffer has to copy from/to device/host buffers
+                CudaUtil.copy(
+                  memBuff,
+                  blockRange.rangeStart,
+                  bounceBuffToUse,
+                  buffOffset,
+                  blockRange.rangeSize())
+              case d: DeviceMemoryBuffer =>
+                memBuff match {
+                  case mh: HostMemoryBuffer =>
+                    // host original => device bounce
+                    d.copyFromHostBufferAsync(buffOffset, mh, blockRange.rangeStart,
+                      blockRange.rangeSize(), serverStream)
+                  case md: DeviceMemoryBuffer =>
+                    // device original => device bounce
+                    d.copyFromDeviceBufferAsync(buffOffset, md, blockRange.rangeStart,
+                      blockRange.rangeSize(), serverStream)
+                  case _ => throw new IllegalStateException("What buffer is this")
+                }
+              case _ => throw new IllegalStateException("What buffer is this")
+            }
+          }
+          buffOffset += blockRange.rangeSize()
+        }
+
+        val alt = AddressLengthTag.from(bounceBuffToUse,
+          blockRanges.head.block.tag)
+
+        alt.resetLength(buffOffset)
+
+        if (windowedBlockIterator.hasNext) {
+          blockRanges = windowedBlockIterator.next()
+        } else {
+          blockRanges = Seq.empty
+          markedDone = true
+        }
+
+        alt
+      } else {
+        null
+      }
+    } catch {
+      case t: Throwable =>
+        logError("Error while copying to bounce buffers on send.", t)
+        throw t
+    }
+
+    logDebug(s"Sending ${buffsToSend} for transfer request, " +
+        s" [peer_executor_id=${transferRequest.executorId()}]")
+
+    buffsToSend
+  }
+
+  /**
+   * Called by the RapidsShuffleServer when it has synchronized with its stream,
+   * allowing us to safely return buffers to the catalog to be potentially freed if spilling.
+   */
+  def releaseAcquiredToCatalog(): Unit = synchronized {
+    require(acquiredBuffs.nonEmpty, "Told to close rapids buffers, but nothing was acquired")
```

Comment:
Again thinking of the close/cleanup when handling an exception at a higher level.  Is it necessary to throw here given it's benign behavior if `acquiredBuffs` is empty?

Reply:
Making this one a warning.